File: Python/233.number-of-digit-one.py
#
# @lc app=leetcode id=233 lang=python3
#
# [233] Number of Digit One
#

# @lc code=start
class Solution:
    def countDigitOne(self, n: int) -> int:
        count = 0
        factor = 1

        while factor <= n:
            higher = n // (factor * 10)
            current = (n // factor) % 10
            lower = n % factor

            if current == 0:
                count += higher * factor
            elif current == 1:
                count += factor * higher + lower + 1
            else:
                count += (higher + 1) * factor

            factor *= 10

        return count
# @lc code=end

File: C++/233.number-of-digit-one.cpp
/*
 * @lc app=leetcode id=233 lang=cpp
 *
 * [233] Number of Digit One
 */

// @lc code=start
class Solution
{
public:
    int countDigitOne(int n)
    {
        int count = 0;
        long long factor = 1;

        while (factor <= n)
        {
            long long higher = n / (factor * 10);
            long long current = (n / factor) % 10;
            long long lower = n % factor;

            if (current == 0)
                count += higher * factor;
            else if (current == 1)
                count += higher * factor + lower + 1;
            else
                count += (higher + 1) * factor;
            factor *= 10;
        }

        return count;
    }
};
// @lc code=end
